### PR TITLE
BUG:datetime list starting with none

### PR DESCRIPTION
**Pull request by MikiPWata**
## PR Summary
#23580 
- cbook/__init__.py
  - modified safe_first_element() to obtain the first non-None element
- test_cbook.py
  - added a test for safe_first_element()
## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ N/A] New features are documented, with examples if plot related.
- [ N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/cbook/__init__.py
Comment (jklymak) [inline lib/matplotlib/cbook/__init__.py:1693]: OK, but now this is _not_ the first element, its the first non-None element.  I'm not sure it makes sense to do this here...

Comment (MikiPWata) [reply lib/matplotlib/cbook/__init__.py:1693]: Thank you for the review. Indeed. Then I guess I have to change where it says "first element" to "first non-None element"

**Review by tacaswell — APPROVED** [review]
This should be squashed merged or squashed and rebased.

**Comment by tacaswell** [comment]
@MikiPWata Do you feel comfortable squashing this to one commit?

**Comment by MikiPWata** [comment]
> @MikiPWata Do you feel comfortable squashing this to one commit?

yes! I will squash them into 1 commit

File: lib/matplotlib/tests/test_cbook.py
Comment (tacaswell) [inline lib/matplotlib/tests/test_cbook.py:894]:
```suggestion
```

Do not actually use `y` in the test.

**Comment by tacaswell** [comment]
Thank you, this is now waiting for a second reviewer to merge.

File: lib/matplotlib/cbook/__init__.py
Comment (jklymak) [reply lib/matplotlib/cbook/__init__.py:1693]: .. and rename the method or wrap the None check in a new private method that is a level up?  This is a public method and this is theoretically a breaking change for someone using it and the new behaviour no longer matches the method name

Comment (MikiPWata) [reply lib/matplotlib/cbook/__init__.py:1693]: @jklymak 
Please correct me if I misunderstood this. I guess I should make a new private method(like _safe_first_non_none_element()), keeping the whole old safe_first_element().
Or, I still need to have the original safe_first_element() even if I rename the method.

Comment (timhoffm) [reply lib/matplotlib/cbook/__init__.py:1693]: I think

`def safe_first_element(obj, skip_none=False):`

is the best way to go here. We should explicitly opt in to skipping None where needed.

Comment (jklymak) [reply lib/matplotlib/cbook/__init__.py:1693]: I think @timhoffm suggestion makes sense.  We could wrap that in a `_safe_first_non_none` or some such for our internal use if that is easier than adding a kwarg to each call...

Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py:1693]: On more consideration, I agree with @timhoffm .  I thought this was an internal helper not a public.

File: lib/matplotlib/cbook/__init__.py
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py]:
```suggestion
def safe_first_element(obj, skip_none=False):
```

**Comment by tacaswell** [comment]
Please squash to one commit again.

Doing this as

```python
    return next(val for val in obj if not (skip_none and val is None))   # I _think_ this logic is correct
```

**Comment by MikiPWata** [comment]
@tacaswell 
oh I commited them by accident.... should I squish them as well?

**Comment by MikiPWata** [comment]
not sure why so many pytests are failing....

**Comment by MikiPWata** [comment]
not sure why many pytests are failing....

**Comment by timhoffm** [comment]
Test failures are likely unrelated https://gitter.im/matplotlib/matplotlib?at=62f2d60b9d3c186299235d4e

File: lib/matplotlib/cbook/__init__.py
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py:1711]:
```suggestion
def _safe_first_non_none(obj, skip_none=True):
```

File: lib/matplotlib/cbook/__init__.py
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py:1687]:
```suggestion
def safe_first_element(obj):
```

File: lib/matplotlib/cbook/__init__.py
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py:1695]: This whole function body can now be

```
return _safe_first_non_none(obj, skip_none=False)
```

**Comment by tacaswell** [comment]
And with the suggested changes we need to find all our usage of `safe_first_element` and replace with `_safe_first_non_none`.

This way we (internally) always use the new behavior, if anyone is using this method for their own reasons they keep the same old behavior, and we only have one copy of the logic!

**Comment by tacaswell** [comment]
I think these test failures are real and caused by the changes in this PR.

File: lib/matplotlib/cbook/__init__.py
Comment (timhoffm) [reply lib/matplotlib/cbook/__init__.py:1711]: Only from mobile, so maybe I’m missing some context, but why do we have two methods now? Can’t we integrate the logic easily in the original method, that now grew the skip_none parameter already? Also, having a method `safe_first_non_none()` that is configurable via `skip_none` seems contradictory in itself.

Comment (MikiPWata) [reply lib/matplotlib/cbook/__init__.py]: @timhoffm 
I created the `_safe_first_non_none()` as a helper function. I guess we need to use kwargs as well  if we are integrating the logic in  `safe_first_element()`?

**Comment by MikiPWata** [comment]
> And with the suggested changes we need to find all our usage of `safe_first_element` and replace with `_safe_first_non_none`.
> 
> This way we (internally) always use the new behavior, if anyone is using this method for their own reasons they keep the same old behavior, and we only have one copy of the logic!

I have also changed all `safe_first_element()` in test files to `_safe_first_none()`.

**Review by tacaswell — APPROVED** [review]
We should also consider deprecating `safe_get_first_element`, but I think we have asked @MikiPWata to do enough!

**Review by jklymak — APPROVED** [review]

**Comment by jklymak** [comment]
Thanks @MikiPWata!

**Pull request merged**
